### ISSUE TITLE
scheduler.step() should be called after each epoch instead of after each batch iteration

### ISSUE THREAD

**Issue by tomatoboy-hub** [rpryzant/causal-bert-pytorch#9]
Currently, `scheduler.step()` is called after each batch iteration during training. 
However, it is more appropriate to call `scheduler.step()` after each epoch to ensure proper learning rate adjustment. I think
``` python
for epoch in range(epochs):
    for step, batch in enumerate(dataloader):
        optimizer.step()
        losses.append(loss.detach().cpu().item())
    scheduler.step()  # <-- Move scheduler step here
```

**Comment by rpryzant** [comment]
Thank you! Hmm this is a good question, I checked around and in this case I think  `get_linear_schedule_with_warmup` should be applied at every step, maybe there are other schedulers for each epoch. 

See e.g. https://github.com/huggingface/transformers/blob/594ca6deadb6bb79451c3093641e3c9e5dcfa446/examples/contrib/run_openai_gpt.py#L252

And the docs which are defined in terms of num steps https://huggingface.co/transformers/v3.0.2/main_classes/optimizer_schedules.html

And code which looks like intended to be adapted for per-step basis https://huggingface.co/transformers/v3.0.2/_modules/transformers/optimization.html#get_linear_schedule_with_warmup
